app/decorators/solidus_zones_by_city/models/address.rb

```diff
@@ -6,7 +6,18 @@ module Address
       extend ActiveSupport::Concern
 
       included do
-        belongs_to :city, class_name: 'Spree::City'
+        belongs_to :locality, class_name: 'Spree::Locality', optional: true
+
+        blacklisted_validation_attributes = %i[city].freeze
```

Comment:
what if we add a before_validation callback to fill address.city with locality.name ? only when it is not present already, and all validations and everything from other extensions will keep working

Reply:
makes sense. I'll update the PR.